doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,416 @@
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  read-only, and T2 is read-only.
+
+#### Strict Serializable
+
+See [https://jepsen.io/consistency/models/strict-serializable](https://jepsen.io/consistency/models/strict-serializable)
+for an in depth formal discussion of Strict Serializable. Strict Serializable is a combination of Serializable and
+Linearizable, which means that the total order formed by all transactions must be constrained by their real time
+occurrences. So if transaction t1 happened in real time before transaction t2, then t1 must be ordered before t2 in the
+total ordering.
+
+Operations that only include user tables are currently Strict Serializable. This is made easy by using a single global
+timestamp and a single thread in the coordinator to order operations. Writes happen at a timestamp larger than all
+previous operations. Reads happen at a timestamp equal to or larger than all previous operations.
+
+Operations that involve reading from views over upstream data are not currently Strict Serializable. It is possible to
+read from a view at one timestamp and then later read from the same view at an earlier timestamp. Consider the following
+scenario:
+
+1. Client reads from view v1 that has an upper of `ts + 1` and a since of `ts - 10`. The read will happen at `ts`.
+2. Client reads from a join over v1 and view v2. v2 has an upper of `ts - 5` and some since less or equal to `ts - 6`.
+   The read will happen at `ts - 6`.
+
+Reference [Timestamp Selection](#Timestamp Selection) to see why these timestamps are selected. This is an obvious
+violation of strict serializability because the second read happens later than the first read in real time, but it is
+assigned an earlier timestamp.
+
+In order to achieve Strict Serializability Materialize must always select a timestamp greater than all previously
+selected timestamps. This will ensure that if an operation happens later in real time, then it will be assigned a later
+timestamp. Consecutive reads can all be assigned an equal timestamp as long as there’s no writes in between them. This
+is made somewhat easy by the fact that all timestamp selection happens on a single thread so no coordination is needed.
+On implementation for this is to consult the global timestamp for all reads, not just ones involving user tables. If
+the `[since, upper)` of any object drifts ahead of the global timestamp, causing the selected timestamp to be greater
+than the global timestamp, then the global timestamp can be fast-forwarded to the selected timestamp.
+
+For the purposes of this document, the approach described above is sufficient to achieve a consistency level of Strict
+Serializability. However, it introduces a performance issue that we will now discuss. Consecutive writes or read/write
+cycles (read followed by write, followed by read, followed by write, etc) will increase the global timestamp in an
```

Comment:
I wanted to include this to scenario anyway because there could be solutions (like a WAL) that only fixes this problem for consecutive write but doesn't fix this problem for consecutive read followed writes.

Reply:
Just removed this sentence.